Fix complete MPU with missing parts

KillianG opened the pull request:
In case of a complete MPU error after the storage of the object MD, the abort must delete object MDs and object Data, some part may have been deleted already in the MPU metadata, hence we need to verify in the object MD for the parts it's linked to, to avoid creating orphans

Issue: CLDSRV-570

bert-e commented:
# Hello killiang,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-570 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.36`


Please check the `Fix Version/s` of CLDSRV-570, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
In case of an abort, delete data and MD of object

In case of a complete MPU error after the storage of the object MD, the abort must delete object MDs and object Data, some part may have been deleted already in the MPU metadata, hence we need to verify in the object MD for the parts it's linked to, to avoid creating orphans

Issue: CLDSRV-570

jonathan-gramain commented on lib/api/apiUtils/object/abortMultipartUpload.js:
```diff
@@ -105,9 +106,39 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
+            }, () => next(null, mpuBucket, storedParts, destBucket, objectMD));
         },
-        function deleteMetadata(mpuBucket, storedParts, destBucket, next) {
+        function deleteObjectData(mpuBucket, storedParts, destBucket, objectMD, next) {
+            // Filtering parts that has already been delete by the previous step
+            let partLocations = storedParts.map(item => item.value.partLocations);
+            partLocations = [].concat(...partLocations);
+            partLocations = new Set(partLocations.map(loc => loc.key));
+            const objectLocationLeft = objectMD.location.filter(loc => !partLocations.has(loc.key));
+
+            return async.eachLimit(objectLocationLeft, 5, (loc, cb) => {
+                data.delete(loc, log, err => {
+                    if (err) {
+                        log.fatal('delete object data failed', { err });
+                    }
+                    cb();
+                });
+            }, () => next(null, mpuBucket, storedParts, destBucket, objectMD));
+        },
+        function deleteObjectMetadata(mpuBucket, storedParts, destBucket, objectMD, next) {
```
If I'm not mistaken I think it's better to delete the metadata first, to avoid leaving the object in an inconsistent state (i.e. unreadable), especially if there are failures.

KillianG replied:
correct me if I'm wrong, but if we crash for some reason after deleting the metadata first, we will end up with orphans never recoverable nor deletable.. ?

williamlardier replied:
From what I understand, usually we prefer having orphans in the storage layer rather than having an orphan in the metadata (orphan in the sense that we cannot do anything with it, and the data is not accessible). Both options are not perfect, but from the client point of view it allows to have a clean state. Nowadays we run tools to cleanup such orphans, but later we may be smarter and queue these operations internally...

Commit:
Remove object MD before deleting data

Issue: CLDSRV-570

Commit:
Add test for abort multipart for the fix of the issue

CLDSRV-570

Commit:
Adding logs to ease future debugging

Issue: CLDSRV-570

jonathan-gramain commented on lib/api/apiUtils/object/abortMultipartUpload.js:
```diff
@@ -105,9 +112,47 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
+            }, () => next(null, mpuBucket, storedParts, destBucket, objectMD));
+        },
+        function deleteObjectMetadata(mpuBucket, storedParts, destBucket, objectMD, next) {
+            if (!objectMD || metadataValMPUparams.uploadId !== objectMD.uploadId) {
```
I don't think it's a huge issue and not sure we can do much better, but we have to keep in mind that in case the bucket is non-versioned, we may end up deleting a newer "version" of the object that overwrote an older "version", which ends with neither the old nor the new state (no object at all, where there were always an object before). However, the application is supposed to want to overwrite the old object in that case so shouldn't care about it, and the application aborts the new object which means it doesn't care about it either. But it could become an issue if e.g. the application needs to read the old object in order to write the new object on top, since with the deletion it will not be able to anymore.

I am thinking out loud here, it may be okay for the time being, but maybe longer term we need to have something more clever to get rid of orphaned MPU keys that doesn't delete parts that are already part of a valid object.

williamlardier replied:
We discussed this within the team and the problem wasn't clear to us, I think I got it, so i'm trying to rephrase- it, please let me know if I'm wrong:

For example, given a non-versioned bucket.

- We create an MPU for `objectA`, complete it: `objectA` exists with upload id `1`.
- Then we do another MPU for the same object (upload id `2`).
- Then the CompleteMPU fails after cloudserver replaced `objectA` with upload id `2`.
- Before the abort, we replace the object again (with a PUT or whatever).
- And then we call the abort, and we end up deleting the wrong object.

But in this very last case, that's what we are actually trying to avoid: if there is no object md, or if the upload id does not match, then we do nothing, so nothing is deleted.
Am I missing something?

francoisferrand replied:
I think @jonathan-gramain is mentioning the case where the user would want to "update the data" (think download/edit/upload) : the first upload fails, but the Abort causes the data to be removed (neither initial not new version are available anymore).... so the application cannot do the update anymore...

Probably not a big issue for now (it is already like that), but ideally we find a solution in the future...

KillianG replied:
Created a ticket to keep track of that issue but won't address it in this PR
https://scality.atlassian.net/browse/CLDSRV-579

jonathan-gramain commented on lib/api/apiUtils/object/abortMultipartUpload.js:
```diff
@@ -105,9 +112,47 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
+            log.info('Object has existing metadata, deleting them', {
+                method: 'abortMultipartUpload',
+                bucketName,
+                objectKey,
```
It could be good to log the `versionId` and `uploadId` as well (but maybe it's already in the log context?)

francoisferrand replied:
unfortunately, I am not sure it is so simple, to cover all situations (versioned, non-versioned and version-suspended) : it probably works for both versioned and non-versioned cases, but i think it may break for the version-suspended case...

Also not sure if the field needs to be decoded/encoded here...

@Kerkesni Can you confirm? There helper functions to cover all cases, right?

welansari replied:
@francoisferrand If i'm not mistaken this should work as is. internal metadata APIs rely on undecoded versionID.

For null versions:
- Objects created in a versioning suspended bucket do not have a versionID.
- In a versioning suspended bucket, putting a version on top of an already created versioned object creates a null version that has an internal versionID.

Both these cases should work with the current implementation.

williamlardier commented on lib/api/apiUtils/object/abortMultipartUpload.js:
```diff
@@ -105,9 +112,47 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
+                return next(null, mpuBucket, storedParts, destBucket);
+            }
+            // Filtering parts that has already been delete by the previous step
+            let partLocations = storedParts.map(item => item.value.partLocations);
+            partLocations = [].concat(...partLocations);
+            partLocations = new Set(partLocations.map(loc => loc.key));
```
This looks under optimised: we map an array, then we deep copy it and recreate a set with another map. We should be able to merge it and create a set from a reduce of the partLocations keys

williamlardier commented on lib/api/apiUtils/object/abortMultipartUpload.js:
```diff
@@ -105,9 +112,47 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
+
+            return async.eachLimit(objectLocationLeft, 5, (loc, cb) => {
+                data.delete(loc, log, err => {
+                    if (err) {
+                        log.fatal('delete object data failed', { err });
+                    }
+                    cb();
```
We do not return any error in the callback yet log a fatal error: fatal is reserved to application crashes. Shall we log an error instead, and return the callback with the error?

KillianG replied:
Not sure we want to do that, return an error would stop the async.eachLimit, not sure we want that (to create more data orphans..)
This is done this way just above in `deleteData` function

francoisferrand commented on lib/api/apiUtils/object/abortMultipartUpload.js:
```diff
@@ -105,9 +112,47 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
+                bucketName,
+                objectKey,
+            });
+            return metadata.deleteObjectMD(bucketName, objectKey, undefined, log, err => {
```
this should be made version-aware : i.e. in versioned bucket, need to specify the versionID of the object MD to delete...

francoisferrand commented on lib/api/apiUtils/object/abortMultipartUpload.js:
```diff
@@ -105,9 +112,47 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
+                return next(err, mpuBucket, storedParts, destBucket, objectMD);
+            });
+        },
+        function deleteObjectData(mpuBucket, storedParts, destBucket, objectMD, next) {
```
similar to the "parts", shouldn't we first delete data, then metadata? otherwise we can get in a bad state, where we don't know of the parts anymore...

this has the added benefit of allowing to remove all parts in a single stage, just combine the MPU parts with the object parts (if any)...

(note: if we decide to first delete the ObjectMD, to avoid having an object pointing to delete data, we need to do it before removing the MPU parts data, as they may be refereced ; and this allows deleting in a single step...)

KillianG replied:
If I understand correctly, I think we prefer having "data orphans" instead of "metadata orphans" because in case of a "metadata orphans" user can think of a data loss because they can list and head the object but not get it..

This has been discussed here with jonathan
https://github.com/scality/cloudserver/pull/5691#discussion_r1841460336

francoisferrand replied:
if we prefer having data orphans, then we need to delete the (object) metadata first : i.e. delete object metadata before the part's data (i.e. before `deleteData()` function), since the parts are likely used in the object as well...

...and doing that we don't need to add a separate step to remove the remaining object's data: we simply need to remove both the parts' data and object data in the `deleteDate()` step.

Commit:
Fix delete object MD with version ID

Issue: CLDSRV-570

Commit:
fix space before brackets

francoisferrand commented on tests/unit/api/multipartUpload.js:
```diff
@@ -30,6 +30,7 @@ const metadataswitch = require('../metadataswitch');
 const { metadata } = storage.metadata.inMemory.metadata;
 const metadataBackend = storage.metadata.inMemory.metastore;
 const { ds } = storage.data.inMemory.datastore;
+const ms = storage.metadata.inMemory.metastore;
```
already defined at line 31: `const metadataBackend = storage.metadata.inMemory.metastore;`

Commit:
Change function order in abortMultipartUpload to avoid code duplication and MD orphans
Remove duplicated import in tests

Issue: CLDSRV-570

francoisferrand commented on lib/api/apiUtils/object/abortMultipartUpload.js:
```diff
@@ -98,6 +126,13 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
             locations = [].concat(...locations);
+
+            if (objectMD?.location) {
+                const objectLocationLeft = objectMD.location.filter(loc =>
```
nit: "left" looks like the directoin, maybe best to rename to something more intuitive, maybe `remainingObjectLocations`

more importantly, this can be quite CPU intensive for large objects (`n^2` complexity, with `n` up to 10000): need to make this more efficient...

welansari replied:
Maybe a more optimized approach for this would be to use a map ?

```
if (objectMD?.location) {
    const locationMap = new Map(locations.map(loc => [loc.key, true]));
    const objectLocationLeft = objectMD.location.filter(loc => !locationMap.has(loc.key));
    locations = locations.concat(objectLocationLeft);
}
```

francoisferrand replied:
since we are building a list of locations, maybe we could directly use a `map` instead (and pass it to async.eachLimit directly, to process each "pair"?)

KillianG replied:
Pushed a more optimized version here
https://github.com/scality/cloudserver/pull/5691/commits/e9ae9c22ec15e3ae2965eef3c3d1115d15a69c2a

FOr locations being a Set directly, I can do it, but as we'll need to map over it anyway for getting keys, we'll end up needing an array (as `map` function does not exist for Sets)

williamlardier commented on lib/api/apiUtils/object/abortMultipartUpload.js:
```diff
@@ -66,26 +71,49 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
+                bucketName,
+                objectKey,
+                uploadId,
+                versionId: objectMD.versionId
```
```suggestion
                versionId: objectMD.versionId,
```

williamlardier commented on lib/api/apiUtils/object/abortMultipartUpload.js:
```diff
@@ -66,26 +71,49 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
+            // In case there has been an error during cleanup after a complete MPU
+            // (e.g. failure to delete MPU MD in shadow bucket),
+            // we need to ensure that the MPU metadata is deleted.
+            log.info('Object has existing metadata, deleting them', {
```
I would put the upload id in the log, if we can have the info, so we can track the problem more easily

KillianG replied:
When we are in this log, both uploadId are the same
https://github.com/scality/cloudserver/pull/5691/files#diff-75b103d02d5ce4b1be1ff37a8a3058d683af6f676c0acef32e4fa1929d23ae7dR96

Commit:
Optimize listing of keys for data deletion

Issue: CLDSRV-570

Commit:
Even better optimize

Issue: CLDSRV-570

KillianG commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.8`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`
* `development/8.7`



Please check the status of the associated issue CLDSRV-570.

Goodbye killiang.




*The following options are set:* **approve**
